---
title: "TreeExp startup guide"
author: Hang Ruan
date: Sep 9th, 2016
output: md_document 
---

# TreeExp
*TreeExp* is an *R* package that performs analyses of expression evolution
from *RNA-seq* data, including optimized input formatting, normalization
and pair-wise distance evaluation, expression character tree inference and
preliminary phylogenetic network analysis.

*TreeExp* package is under active developing, current stable version 1.0 is available at <https://github.com/hr1912/TreeExp>.

A convenient way to install package from github is through *devtools* package:

```{r, eval=FALSE}
install.packages('devtools')
devtools::install_github("hr1912/TreeExp")
```

Users can also download *TreeExp* package and install locally through:

```{r, eval=FALSE}
install.packages("filePath/TreeExp.1.0.tar.gz", repos = NUll, type = "source")
```

Load the package in the usual way:

```{r, warning=FALSE}
library('TreeExp')
```
Expression divergence of two genomes can be considered as concerted evolution of transcriptome from their common ancestor, which can be measured by among orthologous genes' variance components. And more specifically, the expression levels of same tissue across different species can be treated as taxonomic units, which can be placed at the tips of character tree that represents expression evolution of these species.

In here, we will give an example to build a character tree from expression data (expression phylogeny).

### Input Format:

*TreeExp* package takes in reads count data and gene information file in certain format:

1. Gene information file should be a text file in the shape of a matrix, in which values are separated by tabs. `Rows` correspond to orthologous genes and `columns` correspond to species names. And the values in the matrix are in the format of "`GeneId:GeneLength`".

2. Reads count file should also be a text file in the matrix shape, `Rows` correspond to orthologous genes which should be in one-to-one correspondence with rows in Gene information file, though gene ids are displayed in reads count file. `Columns` correspond to sample names. Sample names are in format of "`TaxaName_SubtaxaName_ReplicatesName`". 

The example files are included in the TreeExp package, which can be found in `extdata` folder in the package. One can load them in to take a look:

```{r, warning=FALSE}

readsCount.table = read.table(system.file('extdata/tetraexp.reads.count.raw.txt', 
                                            package='TreeExp'), header = T)
head(readsCount.table[,1:10])

geneInfo.table = read.table(system.file('extdata/tetraexp.gene.length.ortholog.txt',
                                        package='TreeExp'), header = T)
head(geneInfo.table)

```

### Construction:

The construction function `TEconstruct` loads in the reads count data file as well as a gene information file, and wraps them in a list of *taxonExp* objects (one *taxaExp* object).

In the package, we include files transformed from six tissues' expression reads count data of nine tetrapod species.
If you want to transform your own data, a transformation Perl script `format2treeexp.pl` to format raw outputs of *TopHat2* to "*TreeExp* compatible" is available at `tools` folder in the package.
Or you can access the script at
<https://github.com/hr1912/TreeExp/blob/master/tools/format2treeexp.pl>

```{r, eval=FALSE}
taxa.objects = TEconstruct(readsCountFP = system.file('extdata/tetraexp.reads.count.raw.txt', package='TreeExp'),
  geneInfoFP = system.file('extdata/tetraexp.gene.length.ortholog.txt', package='TreeExp'), 
  taxa = "all", subtaxa = c("Brain", "Cerebellum"), calRPKM=TRUE, rmOut=TRUE)
```

The construction process takes **several minutes** on a desktop computer depending on data size and hardware performance. Specify **"taxa"** and **"subtaxa"** options in the function when using partial of your data. The construction process will be faster.
If you are hesitated to test the *TreeExp*, the package has already bundled a constructed object and you can load the object through:

```{r}
data(tetraexp)
```

You can take a look at what the loaded objects:
```{r}
print(tetraexp.objects, details = TRUE)
```

```{r}
print(tetraexp.objects[[1]], printlen = 6)
head(tetraexp.objects[[1]]$rpkm.rmOut)
```

### Distance matrix:

Let us quickly jump to the issue of creating phylogeny from *taxaExp* object. 
First, we generate a distance matrix:

```{r, message=FALSE}
dismat <- expdist(tetraexp.objects, taxa = "all",
                 subtaxa = "Brain",
                 method = "pea")
dismat
```

You can specify **"taxa"** and **"subtaxa"** options in the `expdist` function as well. The default model **"pea"** is to calculate pair-wise distances by Pearson distance, which equals 1-Pearsonâ€™s coefficient of expression level. 

### Expression character tree:

After distance matrix is created, you can construct character tree by Neighbor-Joining, and bootstrap values can also be generated by `boot.exphy` function:

```{r, message=FALSE, warning=FALSE, results='hide'}
tr <- NJ(dismat)
tr <- root(tr, "Chicken_Brain")
bs <- boot.exphy(tr, tetraexp.objects, method = "pea",
                 B = 100, rooted = "Chicken_Brain")
tr$node.label = bs
plot(tr, show.node.label = TRUE)
```

By now, an expression tree is successfully constructed. The tree shows expression patterns' similarities in selected genes of designated species. The expression tree is largely in accordance with species tree with minor discrepancy. 

Phenomenon of evolutionary history dominates the evolutionary expression pattern can be described as phylogenetic signals. One way to interpret highly consistent expression character tree is that expression levels of transcriptome, representing the regulatory changes, accumulated over time. Though not as concrete as sequence data, expression levels generated from transcriptome data across species show strong phylogenetic signals.
