<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Hang Ruan" />


<title>TreeExp basics</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">TreeExp basics</h1>
<h4 class="author"><em>Hang Ruan</em></h4>



<p><em>TreeExp</em> can be loaded the package in the usual way:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">'TreeExp'</span>)</code></pre></div>
<div id="input-format" class="section level3">
<h3>Input Format:</h3>
<p><em>TreeExp</em> package takes in reads count data and gene information file in certain format:</p>
<ol style="list-style-type: decimal">
<li><p>Gene information file should be a text file in the shape of a matrix, in which values are separated by tabs. <code>Rows</code> correspond to orthologous genes and <code>columns</code> correspond to species names. And the values in the matrix are in the format of “<code>GeneId:GeneLength</code>”.</p></li>
<li><p>Reads count file should also be a text file in the matrix shape, <code>Rows</code> correspond to orthologous genes which should be in one-to-one correspondence with rows in Gene information file, though gene ids are displayed in reads count file. <code>Columns</code> correspond to sample names. Sample names are in format of “<code>TaxaName_SubtaxaName_ReplicatesName</code>”.</p></li>
</ol>
<p>The example files are included in the TreeExp package, which can be found in <code>extdata</code> folder in the package. One can load them in to take a look:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">readsCount.table =<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">system.file</span>(<span class="st">'extdata/tetraexp.read.counts.raw.txt'</span>, 
                                            <span class="dt">package=</span><span class="st">'TreeExp'</span>), <span class="dt">header =</span> T)
<span class="kw">head</span>(readsCount.table[,<span class="dv">1</span>:<span class="dv">10</span>])</code></pre></div>
<pre><code>##   homoSapienGeneId Human_Brain_Female Human_Brain_Male1 Human_Brain_Male2
## 1  ENSG00000198824              16323             11147             19507
## 2  ENSG00000118402              31883             19242             32321
## 3  ENSG00000166167             102711             80104            141338
## 4  ENSG00000144724              51020             37861             43906
## 5  ENSG00000183508                988              1443               760
## 6  ENSG00000008086              27839             10208             20666
##   Human_Brain_Male3 Human_Brain_Male4 Human_Brain_Male5
## 1              4402              6521               683
## 2              9841              1164              1362
## 3             59947             23104             14613
## 4             24645              3968              5653
## 5               456               227               456
## 6             12970              2569              1940
##   Human_Cerebellum_Female Human_Cerebellum_Male Human_Heart_Female
## 1                   20032                 43008               3381
## 2                   26732                 57081                228
## 3                   56103                105866              17022
## 4                   26365                 66177              41876
## 5                     988                  1215              16913
## 6                    4987                 11862                878</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">geneInfo.table =<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">system.file</span>(<span class="st">'extdata/tetraexp.length.ortholog.txt'</span>,
                                        <span class="dt">package=</span><span class="st">'TreeExp'</span>), <span class="dt">header =</span> T)
<span class="kw">head</span>(geneInfo.table)</code></pre></div>
<pre><code>##                  Human              Chimpanzee                  Bonobo
## 1 ENSG00000198824:3788 ENSPTRG00000023033:3786 ENSPTRG00000023033:3786
## 2 ENSG00000118402:3042 ENSPTRG00000018370:2962 ENSPTRG00000018370:2962
## 3 ENSG00000166167:6255 ENSPTRG00000002869:6153 ENSPTRG00000002869:6153
## 4 ENSG00000144724:9524 ENSPTRG00000015067:6726 ENSPTRG00000015067:6726
## 5 ENSG00000183508:5751 ENSPTRG00000001166:5717 ENSPTRG00000001166:5717
## 6 ENSG00000008086:3576 ENSPTRG00000021711:2874 ENSPTRG00000021711:2874
##                   Gorilla               Orangutan                 Macaque
## 1 ENSGGOG00000009918:3808 ENSPPYG00000005537:2442 ENSMMUG00000023599:3781
## 2 ENSGGOG00000009359:2890  ENSPPYG00000016794:945 ENSMMUG00000020208:2960
## 3 ENSGGOG00000005023:5882 ENSPPYG00000002583:6130 ENSMMUG00000006741:6001
## 4 ENSGGOG00000006858:6049 ENSPPYG00000013752:4933 ENSMMUG00000012487:5949
## 5 ENSGGOG00000004149:5727 ENSPPYG00000000972:1987 ENSMMUG00000004907:3522
## 6 ENSGGOG00000010796:3354 ENSPPYG00000020166:3024 ENSMMUG00000005063:3077
##                     Mouse                 Opossum                Platypus
## 1 ENSMUSG00000047710:4040 ENSMODG00000003128:2568 ENSOANG00000000610:2487
## 2 ENSMUSG00000032262:2146  ENSMODG00000018420:945  ENSOANG00000001290:909
## 3 ENSMUSG00000025217:2984 ENSMODG00000011788:1888 ENSOANG00000007316:1936
## 4 ENSMUSG00000021745:9375 ENSMODG00000002722:4448 ENSOANG00000006011:3968
## 5 ENSMUSG00000044468:5640 ENSMODG00000023227:1194 ENSOANG00000003065:1335
## 6 ENSMUSG00000031292:3484 ENSMODG00000017140:2589 ENSOANG00000004037:2607
##                   Chicken
## 1 ENSGALG00000016813:2412
## 2 ENSGALG00000015876:1322
## 3 ENSGALG00000007820:1962
## 4 ENSGALG00000007177:5396
## 5 ENSGALG00000014453:1179
## 6 ENSGALG00000016529:2712</code></pre>
</div>
<div id="construction" class="section level3">
<h3>Construction:</h3>
<p>The construction function <code>TEconstruct</code> loads in the reads count data file as well as a gene information file, and wraps them in a list of <em>taxonExp</em> objects (one <em>taxaExp</em> object).</p>
<p>In the package, we include files transformed from six tissues’ expression reads count data of nine tetrapod species. If you want to transform your own data, a transformation Perl script <code>format2treeexp.pl</code> to format raw outputs of <em>TopHat2</em> to “<em>TreeExp</em> compatible” is available at <code>tools</code> folder in the package. Or you can access the script at <a href="https://github.com/hr1912/TreeExp/blob/master/tools/format2treeexp.pl" class="uri">https://github.com/hr1912/TreeExp/blob/master/tools/format2treeexp.pl</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">taxa.objects =<span class="st"> </span><span class="kw">TEconstruct</span>(<span class="dt">readCountsFP =</span> <span class="kw">system.file</span>(<span class="st">'extdata/tetraexp.read.counts.raw.txt'</span>, <span class="dt">package=</span><span class="st">'TreeExp'</span>),
  <span class="dt">geneInfoFP =</span> <span class="kw">system.file</span>(<span class="st">'extdata/tetraexp.length.ortholog.txt'</span>, <span class="dt">package=</span><span class="st">'TreeExp'</span>), 
  <span class="dt">taxa =</span> <span class="st">&quot;all&quot;</span>, <span class="dt">subtaxa =</span> <span class="kw">c</span>(<span class="st">&quot;Brain&quot;</span>, <span class="st">&quot;Cerebellum&quot;</span>), <span class="dt">normalize =</span> <span class="st">&quot;TPM&quot;</span>)</code></pre></div>
<p>The construction process takes <strong>several minutes</strong> on a desktop computer depending on data size and hardware performance. Specify <strong>“taxa”</strong> and <strong>“subtaxa”</strong> options in the function when using partial of your data. The construction process will be faster. If you are hesitated to test the <em>TreeExp</em>, the package has already bundled a constructed object and you can load the object through:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(tetraexp)</code></pre></div>
<p>You can take a look at what the loaded objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(tetraexp.objects, <span class="dt">details =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## 
##  59 taxonExp objects 
## 
## object 1 : Human      Brain 
## object 2 : Human      Cerebellum 
## object 3 : Human      Heart 
## object 4 : Human      Kidney 
## object 5 : Human      Liver 
## object 6 : Human      Testis 
## object 7 : Chimpanzee     Brain 
## object 8 : Chimpanzee     Cerebellum 
## object 9 : Chimpanzee     Heart 
## object 10 : Chimpanzee    Kidney 
## object 11 : Chimpanzee    Liver 
## object 12 : Chimpanzee    Testis 
## object 13 : Bonobo    Brain 
## object 14 : Bonobo    Cerebellum 
## object 15 : Bonobo    Heart 
## object 16 : Bonobo    Kidney 
## object 17 : Bonobo    Liver 
## object 18 : Bonobo    Testis 
## object 19 : Gorilla   Brain 
## object 20 : Gorilla   Cerebellum 
## object 21 : Gorilla   Heart 
## object 22 : Gorilla   Kidney 
## object 23 : Gorilla   Liver 
## object 24 : Gorilla   Testis 
## object 25 : Orangutan     Brain 
## object 26 : Orangutan     Cerebellum 
## object 27 : Orangutan     Heart 
## object 28 : Orangutan     Kidney 
## object 29 : Orangutan     Liver 
## object 30 : Macaque   Brain 
## object 31 : Macaque   Cerebellum 
## object 32 : Macaque   Heart 
## object 33 : Macaque   Kidney 
## object 34 : Macaque   Liver 
## object 35 : Macaque   Testis 
## object 36 : Mouse     Brain 
## object 37 : Mouse     Cerebellum 
## object 38 : Mouse     Heart 
## object 39 : Mouse     Kidney 
## object 40 : Mouse     Liver 
## object 41 : Mouse     Testis 
## object 42 : Opossum   Brain 
## object 43 : Opossum   Cerebellum 
## object 44 : Opossum   Heart 
## object 45 : Opossum   Kidney 
## object 46 : Opossum   Liver 
## object 47 : Opossum   Testis 
## object 48 : Platypus      Brain 
## object 49 : Platypus      Cerebellum 
## object 50 : Platypus      Heart 
## object 51 : Platypus      Kidney 
## object 52 : Platypus      Liver 
## object 53 : Platypus      Testis 
## object 54 : Chicken   Brain 
## object 55 : Chicken   Cerebellum 
## object 56 : Chicken   Heart 
## object 57 : Chicken   Kidney 
## object 58 : Chicken   Liver 
## object 59 : Chicken   Testis</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(tetraexp.objects[[<span class="dv">1</span>]], <span class="dt">printlen =</span> <span class="dv">6</span>)</code></pre></div>
<pre><code>## 
## One taxonExp object
## Taxon name:  Human 
## Subtaxon name:  Brain 
## Total gene number:  5636 
## Total bio replicates number:  6 
## Bio replicates titles:
## [1] &quot;Female&quot; &quot;Male1&quot;  &quot;Male2&quot;  &quot;Male3&quot;  &quot;Male4&quot;  &quot;Male5&quot; 
## Outliers NOT removed
## Normalized expression value calculated
## Normalized method:  TPM 
## Over-dispersion parameter omega NOT calculated</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(tetraexp.objects[[<span class="dv">1</span>]]$normExp.val)</code></pre></div>
<pre><code>##                 Human_Brain_Female Human_Brain_Male1 Human_Brain_Male2
## ENSG00000198824          28.181700         22.882260          32.29216
## ENSG00000118402          68.545215         49.186043          66.62576
## ENSG00000166167         107.390735         99.581296         141.69295
## ENSG00000144724          35.034687         30.911749          28.90820
## ENSG00000183508           1.123546          1.951074           0.82868
## ENSG00000008086          50.913540         22.196988          36.23894
##                 Human_Brain_Male3 Human_Brain_Male4 Human_Brain_Male5
## ENSG00000198824        11.2300374        23.9962144          7.981515
## ENSG00000118402        31.2623165         5.3337453         19.819491
## ENSG00000166167        92.6149878        51.4870922        103.415744
## ENSG00000144724        25.0063652         5.8075209         26.274483
## ENSG00000183508         0.7662361         0.5502005          3.509911
## ENSG00000008086        35.0496431        10.0139426         24.014791</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
